ls1intum/Artemis
# `Development`: Disable logging of sensitive data in e2e cypress tests

### Motivation and Context


We want to connect the Cypress runs on Bamboo and Github with the Cypress Dashboard to allow easier investigation of test failures especially for developers not familiar with the tests. The problem is that the authentication request logs the login credentials and that the Cypress Dashboard is publicly accessible. This results in a leakage of credentials.

### Description

In this PR I disabled the logging of authentication requests and type command, which enter credentials in the UI. Unfortunately simply disabling the logging of the authentication request is not enough because Cypress still prints the response if it considers the request as failed (per default a HTTP status not being 2xx or 3xx is considered a failure). I disabled the default failure assertion for the authentication request, so that it always passes. Since we still want the tests to fail if the authentication request failed I added an assertion which makes sure that the login response has status 200. If that assertion fails it only leaks the status code of the response without any credentials.
I also replaced the custom `loginWithGUI` Cypress command with a LoginPageObject because a custom Cypress command is unnecessary and not consistent with the other tests. I also removed the custom `logout` command because it is not really used and is redundant (it clears the session storage, but this is automatically done by Cypress between tests).
I also fixed a few flaky and failing tests in this PR (refactored footer tests, fixed flaky programming exercise management part)

### Steps for Testing

Run the test suite locally against a test server or local setup and try to find locations where the credentials might be leaked. The credentials are only accessed in the `cy.login` command and in the `Authentication.spec.ts` when entering the credentials in the UI. [The confluence page explains how to setup the Cypress tests](https://confluence.ase.in.tum.de/display/ArTEMiS/Cypress+Tests).

I tried to cause failures in the authentication request to force Cypress to log the error:
Network error in Cypress Test Runner:

<img width="369" alt="OnNetworkError" src="https://user-images.githubusercontent.com/12861668/138160544-01fea24a-eefd-473b-9cbb-5ecc49038f78.PNG">

Network error in console:

<img width="598" alt="OnNetworkErrorConsole" src="https://user-images.githubusercontent.com/12861668/138160650-d164455a-a60b-44d2-b508-03bf5d7e14fe.PNG">

502 Bad Gateway (Test server is starting) in Cypress Test Runner:

<img width="376" alt="TestRunnerOn502" src="https://user-images.githubusercontent.com/12861668/138160804-5ce1cac3-d6a9-4301-b84a-4b31c37c7f9f.PNG">

502 Bad Gateway (Test server is starting) in console:

<img width="608" alt="ConsoleRequestFailure" src="https://user-images.githubusercontent.com/12861668/138160857-6caef93c-0da7-4a78-b275-29522b9bac51.PNG">

As you can see the credentials in the authentication request are not logged anymore.

### Review Progress





#### Code Review
